### PR TITLE
WIP Recompose inputs

### PR DESCRIPTION
**Pull request by chrisnojima**
This PR is to showcase some stuff we can do with recompose to help with dealing with uncontrolled inputs

### CODE REVIEW

File: shared/common-adapters/icon.desktop.js [hunk 1]
```diff
@@ -9,61 +9,59 @@ import {resolveImageAsURL} from '../../desktop/resolve-root'
 import type {Exact} from '../constants/types/more'
 import type {Props} from './icon'
 
-class Icon extends PureComponent<void, Exact<Props>, void> {
-  render () {
-    let color = shared.defaultColor(this.props.type)
-    let hoverColor = shared.defaultHoverColor(this.props.type)
-    let iconType = shared.typeToIconMapper(this.props.type)
+const Icon = (props: Exact<Props>) => {
```
Comment (chrisnojima) [inline shared/common-adapters/icon.desktop.js]: This is me just turning this into a stateless component

File: shared/profile/edit-profile/container.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+type OwnProps = {
+  routeState: {
+    bio: ?string,
+    location: ?string,
+    fullname: ?string,
+  },
+}
+
+export default compose(
+  connect(
```
Comment (chrisnojima) [inline shared/profile/edit-profile/container.js]: Using withState to create local cached values for the profile metadata

Comment (chrisnojima) [inline shared/profile/edit-profile/container.js]: also using the container file naming scheme

File: shared/profile/prove-enter-username-container.js [hunk 3]
```diff
@@ -1,58 +1,48 @@
...
-        } else if (profile.platform === 'zcash') {
-          dispatch(submitZcashAddress())
-        } else {
-          dispatch(submitUsername())
-        }
-      },
-      platform: profile.platform,
-      username: profile.username,
-      waiting: profile.waiting,
+export default compose(
```
Comment (chrisnojima) [inline shared/profile/prove-enter-username-container.js]: Using withState to eliminate the class wrapper

File: shared/search/user-search/search-bar.desktop.js [hunk 4]
```diff
@@ -1,53 +1,49 @@
...
+  onClickService: ServiceFn,
+  onMouseEnter: () => void,
+  onMouseLeave: () => void,
+  onClick: () => void,
+  showingTooltip: boolean,
 }
 
-class SearchBar extends Component<void, Props, void> {
-  _onDebouncedSearch: (overridePlatform?: SearchPlatforms) => void;
+const ServiceIcon =
```
Comment (chrisnojima) [inline shared/search/user-search/search-bar.desktop.js]: Using withHandlers to eliminate closures that change on new props

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready to look at

File: shared/profile/edit-profile/container.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      onCancel: () => dispatch(navigateUp()),
+      onBack: () => dispatch(navigateUp()),
+      onSubmit: (bio, fullname, location) => dispatch(editProfile(bio, fullname, location)),
+    })
+  ),
+  withState('fullname', 'onFullnameChange', props => props.fullname),
+  withState('location', 'onLocationChange', props => props.location),
+  withState('bio', 'onBioChange', props => props.bio),
+  withProps(props => ({
+    onSubmit: () => props.onSubmit(props.fullname, props.location, props.bio),
```
Comment (chromakode) [inline shared/profile/edit-profile/container.js]: Why is this necessary? Can't it be part of the `connect()` above? Shouldn't this use the state instead of props?

File: shared/profile/edit-profile/container.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  })),
+  lifecycle({
```
Comment (chromakode) [inline shared/profile/edit-profile/container.js]: :sheep: This use of recompose seems a bit gratuitous :slightly_smiling_face: I like `withState` a lot because it pulls out pure components, but this feels like a little too much stacking of HOCs. However, if we wrapped the `withState` and `lifecycle` stuff into a HOC for linking state to `routeState` then I think it'd be a win.

File: shared/profile/prove-enter-username-container.js (hunk 3 above)
Comment (chromakode) [inline shared/profile/prove-enter-username-container.js]: Win! :+1:

File: shared/search/user-search/search-bar.desktop.js [hunk 7]
```diff
@@ -1,53 +1,49 @@
...
+  compose(
+    withState('showingTooltip', 'setShowingTooltip'),
+    withHandlers({
+      onClick: props => event => props.onClickService(props.serviceName),
+      onMouseEnter: props => event => props.setShowingTooltip(true),
+      onMouseLeave: props => event => props.setShowingTooltip(false),
+    })
+  )(({serviceName, tooltip, iconType, selected, onClick, onMouseEnter, onMouseLeave, showingTooltip}: ServiceIconProps) => (
```
Comment (chromakode) [inline shared/search/user-search/search-bar.desktop.js]: :sheep: I think the render function should be broken out of this big expression. It's getting a bit unwieldy w/ `)(({`.

File: shared/search/user-search/search-bar.desktop.js [hunk 8]
```diff
@@ -1,53 +1,49 @@
...
+    </Box>
+  )
+)
 
-  constructor (props: Props) {
-    super(props)
-    this._onDebouncedSearch = _.debounce(this._onSearch, 500)
-  }
+class SearchBar extends Component<void, Props, void> {
+  _onDebouncedSearch = _.debounce(this._onSearch, 500)
```
Comment (chromakode) [inline shared/search/user-search/search-bar.desktop.js]: I can't remember, does `this` actually work here? Can this be done with `withHandlers`?

File: shared/search/user-search/search-bar.desktop.js [hunk 9]
```diff
@@ -69,18 +65,18 @@ class SearchBar extends Component<void, Props, void> {
     }
   }
 
-  render () {
-    const services = ['Keybase', 'Twitter', 'Facebook', 'Github', 'Coinbase', 'Reddit', 'Hackernews']
-    const tooltips: {[key: string]: ?string} = {'Hackernews': 'Hacker News'}
+  _services = ['Keybase', 'Twitter', 'Facebook', 'Github', 'Coinbase', 'Reddit', 'Hackernews']
+  _tooltips: {[key: string]: ?string} = {'Hackernews': 'Hacker News'}
```
Comment (chromakode) [inline shared/search/user-search/search-bar.desktop.js]: :sheep: Could move these out of the class entirely if they're not reused elsewhere.

**Comment by chrisnojima** [comment]
i'm changing this up slightly, ignore for now

**Comment by chrisnojima** [comment]
closing this in favor of doing this in smaller pieces and needing to fixup how search works first
